File: 2409-count-days-spent-together/2409-count-days-spent-together.js
/**
 * @param {string} arriveAlice
 * @param {string} leaveAlice
 * @param {string} arriveBob
 * @param {string} leaveBob
 * @return {number}
 */
var countDaysTogether = function(arriveAlice, leaveAlice, arriveBob, leaveBob) {
  const arrive = Math.max(dayOfYear(arriveAlice), dayOfYear(arriveBob));
  const leave = Math.min(dayOfYear(leaveAlice), dayOfYear(leaveBob));
  
  return Math.max(0, leave - arrive + 1);
};

const dayOfYear = (dates) => {
  const days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  const preSum = days.reduce((acc, curr) => (acc.push(acc.at(-1) + curr), acc), [0]);
  const [m, d] = dates.split('-');
  
  return preSum[+m - 1] + +d;
}

File: 2409-count-days-spent-together/2409-count-days-spent-together.py
class Solution:
    def countDaysTogether(self, arriveAlice: str, leaveAlice: str, arriveBob: str, leaveBob: str) -> int:
        def dayOfYear(dates):
            days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
            m, d = dates.split('-')
            return sum(days[:int(m)-1]) + int(d)
          
          
        arrive = max(dayOfYear(arriveAlice), dayOfYear(arriveBob))
        leave = min(dayOfYear(leaveAlice), dayOfYear(leaveBob))
        return max(0, leave - arrive + 1)